### ISSUE TITLE
Argument declaration highlighting blocks Emacs in certain cases

### ISSUE THREAD

**Issue by zzamboni** [ALSchwalm/elvish-mode#10]
I have found at least one case in which `elvish-argument-declaration-pattern` makes Emacs go into an infinite loop (or something like that). For an example, try loading this file: https://github.com/zzamboni/vcsh_elvish/blob/master/.elvish/lib/theme/chain.elv

Emacs' CPU usage goes way up and it hangs. Commenting out these lines fixes the problem:

```
    (,elvish-argument-declaration-pattern . ((1 font-lock-variable-name-face)
                                             (2 font-lock-variable-name-face nil lax)))
```

I have not had a change to dive into this, just letting you know in case you have a chance to look into it.

**Comment by ALSchwalm** [comment]
Hmm, it looks like this is caused emacs doing something strange when it can't find the `]{` at the end of that macro. Things slow down when typing:

    `[asdf asdf asdf asdf asdf asdf`

But not if you type `[]{}` and add args in the middle of the square brackets. Not sure how to fix this just yet.

**Comment by ALSchwalm** [comment]
This should be fixed in 9cf31b453ac79b011b84e83ca11c964c4e647649.

**Comment by zzamboni** [comment]
I can confirm that at least in my test case, this is fixed. Thanks!
